ArmPkg/Drivers/SmmuDxe/SmmuV3Util.c

```diff
@@ -0,0 +1,649 @@
+
+  @param [in]  SmmuBase  The base address of the SMMU.
+
+  @retval EFI_SUCCESS    Success.
+  @retval Others         Failure.
+**/
+EFI_STATUS
+EFIAPI
+SmmuV3DisableTranslation (
+  IN UINT64  SmmuBase
+  )
+{
+  SMMUV3_CR0  Cr0;
+  EFI_STATUS  Status;
+
+  Cr0.AsUINT32 = SmmuV3ReadRegister32 (SmmuBase, SMMU_CR0);
+  if ((Cr0.AsUINT32 & SMMUV3_CR0_SMMU_CMDQ_EVTQ_PRIQ_EN_MASK) != 0) {
+    Cr0.AsUINT32 = Cr0.AsUINT32 & ~SMMUV3_CR0_SMMU_CMDQ_EVTQ_PRIQ_EN_MASK;
+    SmmuV3WriteRegister32 (SmmuBase, SMMU_CR0, Cr0.AsUINT32);
+    Status = SmmuV3Poll (SmmuBase + SMMU_CR0ACK, SMMUV3_CR0_SMMU_CMDQ_EVTQ_PRIQ_EN_MASK, 0);
+    if (Status != EFI_SUCCESS) {
+      DEBUG ((DEBUG_ERROR, "Error SmmuV3Poll: 0x%lx\n", SmmuBase + SMMU_CR0ACK));
+      return Status;
+    }
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Initialise the SMMUv3 to set it in ABORT mode and stop DMA.
+
+  @param [in]  SmmuReg    Base address of the SMMUv3.
+
+  @retval EFI_SUCCESS     Success.
+  @retval EFI_TIMEOUT     Timeout.
+**/
+EFI_STATUS
+EFIAPI
+SmmuV3GlobalAbort (
+  IN  UINT64  SmmuBase
+  )
+{
+  EFI_STATUS  Status;
+  UINT32      RegVal;
+
+  // Attribute update has completed when SMMU_(S)_GBPA.Update bit is 0.
+  Status = SmmuV3Poll (SmmuBase + SMMU_GBPA, SMMU_GBPA_UPDATE, 0);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  // SMMU_(S)_CR0 resets to zero with all streams bypassing the SMMU,
+  // so just abort all incoming transactions.
+  RegVal = MmioRead32 (SmmuBase + SMMU_GBPA);
+
+  // Set the SMMU_GBPA.ABORT and SMMU_GBPA.UPDATE.
+  RegVal |= (SMMU_GBPA_ABORT | SMMU_GBPA_UPDATE);
+
+  MmioWrite32 (SmmuBase + SMMU_GBPA, RegVal);
+
+  // Attribute update has completed when SMMU_(S)_GBPA.Update bit is 0.
+  Status = SmmuV3Poll (SmmuBase + SMMU_GBPA, SMMU_GBPA_UPDATE, 0);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  // Sanity check to see if abort is set
+  Status = SmmuV3Poll (SmmuBase + SMMU_GBPA, SMMU_GBPA_ABORT, SMMU_GBPA_ABORT);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Initialise the SMMUv3 to set Non-secure streams to bypass the SMMU.
+
+  @param [in]  SmmuReg    Base address of the SMMUv3.
+
+  @retval EFI_SUCCESS     Success.
+  @retval EFI_TIMEOUT     Timeout.
+**/
+EFI_STATUS
+EFIAPI
+SmmuV3SetGlobalBypass (
+  IN UINT64  SmmuBase
+  )
+{
+  EFI_STATUS  Status;
+  UINT32      RegVal;
+
+  // Attribute update has completed when SMMU_(S)_GBPA.Update bit is 0.
+  Status = SmmuV3Poll (SmmuBase + SMMU_GBPA, SMMU_GBPA_UPDATE, 0);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  // SMMU_(S)_CR0 resets to zero with all streams bypassing the SMMU
+  RegVal = MmioRead32 (SmmuBase + SMMU_GBPA);
+
+  // TF-A configures the SMMUv3 to abort all incoming transactions.
+  // Clear the SMMU_GBPA.ABORT to allow Non-secure streams to bypass
+  // the SMMU.
+  RegVal &= ~SMMU_GBPA_ABORT;
+  RegVal |= SMMU_GBPA_UPDATE;
+
+  MmioWrite32 (SmmuBase + SMMU_GBPA, RegVal);

```

Comment:
nit: there is a helper func for this and the read above, right?

Reply:
fixed